[3D] Use QGIS material for the terrain

ptitjano opened the pull request:
## Description

This is factored out from https://github.com/qgis/QGIS/pull/57899

It allows to use QGIS material for the terrain instead of native qt3D ones. It will make it possible to customize those shaders. There are 3 different cases to handle: 
- no texture: there is a drop-in replacement with `QgsPhongMaterialSettings`
- texture, no shading: `QgsPhongTexturedMaterialSettings` is extended to handle both textures and textures from a path
- texture, shading. A new material is introduced: `QgsTextureMaterial`

This PR needs https://github.com/qgis/QGIS/pull/58114 to be merged first and thus it includes it.

ptitjano committed:
3d/material: Introduce QgsTextureMaterial

This is the same material as the qt3d one:
`Qt3DExtras::QTextureMaterial`. It will also be used in the following
commit by the terrain.

ptitjano committed:
terrain: Use QgsTextureMaterial for non shading texture

There is no functional change. With this change, the terrain uses a
material defined by QGIS instead of a default Qt3D. This will make it
easier to customize the shader in the future.

ptitjano committed:
terrain: Use PhongMaterialSettings material for the non texture case

There is no functional change. With this change, the terrain uses a
material defined by QGIS instead of a default Qt3D. This will make it
easier to customize the shader in the future.

ptitjano committed:
terrain: Use QgsPhongTexturedMaterial for shading texture

There is no functional change. With this change, the terrain uses a
material defined by QGIS instead of a default Qt3D. This will make it
easier to customize the shader in the future.

ptitjano committed:
qgsphongtexturedmaterialsettings: Do proper opacity cast in toMaterial

nyalldawson commented on python/3d/auto_generated/materials/qgsphongmaterialsettings.sip.in:
```diff
@@ -60,12 +60,12 @@ Returns diffuse color component
 Returns specular color component
 %End
-    double shininess() const;
+    float shininess() const;
```
This was deliberate -- it was to avoid the user-set values changing for them when saving/restoring projects. We can safely convert to float when building the material, but doubles should be used for storing the value to avoid this behaviour.

nyalldawson replied:
> I guess that the textured case should use double

Yes please!

ptitjano replied:
Done.

nyalldawson commented on python/3d/auto_generated/materials/qgsphongtexturedmaterialsettings.sip.in:
```diff
@@ -65,11 +72,11 @@ Returns shininess of the surface
 
 
-    QString diffuseTexturePath() const;
+    QVariant diffuseTexture() const;
```
This is an api break -- can you leave the existing getters/setters, but deprecate them?

ptitjano replied:
This is indeed an api break but the header contains the following sentence: 

```
\warning This is not considered stable API, and may change in future QGIS releases. It is
 * exposed to the Python bindings as a tech preview only.
 ```

nyalldawson replied:
That's our "get out of jail" card for extreme cases only though -- wherever possible we should still be "nice" to plugin developers and avoid breaking API

ptitjano replied:
I don't mind deprecate the existing function but it does look like we some reliable rules to break it or not. 

I remember on a previous PR, you specifically asked me to break the API of some 3D enum. At the time, your argument was: "no one is really using that API." I could use the same argument here: I really don't think that anyone is using that API.

ptitjano replied:
The API is now preserved and the previous functions are deprecated.

nyalldawson commented on python/3d/auto_generated/materials/qgsphongtexturedmaterialsettings.sip.in:
```diff
@@ -65,11 +72,11 @@ Returns shininess of the surface
+    QVariant diffuseTexture() const;
 %Docstring
-Returns the diffuse texture path.
+Returns the diffuse texture component of the material. This can be either a path as a QString or a texture.
```
Can you make a note as to what value types are acceptable for a texture? Eg is a QImage acceptable?

ptitjano replied:
Done.

nyalldawson commented on src/3d/materials/qgstexturematerial.h:
```diff
@@ -0,0 +1,82 @@
+
+/**
+ * \ingroup 3d
+ * \brief A texture material which discard transparent components.
```
Maybe:

```suggestion
 * \brief A unlit texture material which discard transparent components.
```

ptitjano replied:
Done.

nyalldawson commented on tests/src/3d/testqgs3drendering.cpp:
```diff
@@ -485,6 +487,49 @@ void TestQgs3DRendering::testPhongShading()
+
+  delete scene;
+  delete map;
+  QGSVERIFYIMAGECHECK( "polygon3d_extrusion_textured_phong", "polygon3d_extrusion_textured_phong", img, QString(), 40, QSize( 0, 0 ), 2 );
```
These test images look quite detailed, and I'm concerned they'll be tricky to make reliable. Could you increase the texture size to help alleviate this?

ptitjano replied:
The test (qt5 and qt6) works on my machine. The qt6 version works on the ci but not the qt5 version. 
I tried to zoom in a lot to verify that the texture is correct. I have updated the test to zoom out a bit.

ptitjano replied:
Not sure what happens. I'm positive this has nothing to do with the current PR. I got the same result on master. I have dropped this commit and created a separate PR to add the unit test: https://github.com/qgis/QGIS/pull/58188

ptitjano committed:
qgsphongtexturedmaterialsettings: Use a double to store opacity

This is already waht QgsPhongTexturedMaterialSettings does. The reason
for this change is to avoid the user-set values changing for them when
saving/restoring projects.

ptitjano committed:
qgsphongtexturedmaterialsettings: Use a double to store shininess

This is already waht QgsPhongTexturedMaterialSettings does. The reason
for this change is to avoid the user-set values changing for them when
saving/restoring projects.

ptitjano committed:
qgsphongmaterialsettings: Fix setShininess signature

`mShininess` is a double.

nyalldawson commented on src/3d/materials/qgsphongtexturedmaterialsettings.cpp:
```diff
@@ -86,7 +94,10 @@ void QgsPhongTexturedMaterialSettings::writeXml( QDomElement &elem, const QgsRea
   elem.setAttribute( QStringLiteral( "shininess" ), mShininess );
   elem.setAttribute( QStringLiteral( "opacity" ), mOpacity );
-  elem.setAttribute( QStringLiteral( "diffuse_texture_path" ), mDiffuseTexturePath );
+  if ( !mDiffuseTexture.value<QString>().isNull() )
```
```suggestion
  if ( !mDiffuseTexture.value<QString>().isEmpty() ) )
```

nyalldawson commented on src/3d/materials/qgsphongtexturedmaterialsettings.cpp:
```diff
@@ -202,9 +232,35 @@ void QgsPhongTexturedMaterialSettings::addParametersToEffect( Qt3DRender::QEffec
+{
+  if ( mDiffuseTexture.value<Qt3DRender::QTexture2D *>() )
+  {
+    delete mDiffuseTexture.value<Qt3DRender::QTexture2D *>();
```
This looks messy. It's deleted here, but not in the destructor? And what about when the settings are copied/cloned? There's also no mention of lifetime or ownership in the docs. 

Could you add a smart pointer to hold the actual object instead of relying on the raw pointer stored in a variant?

ptitjano replied:
I was completely wrong. When the `diffuse` component is set from a `Qt3DExtras::QTextureMaterial`, the underlying `QImage` needs to be copied because the material takes ownership of the associated parameter. It worked before the QTexture2D from the settings was not deleted..

Now: 
- the `QImage` is correctly copied and used to create a texture parameter used by the material in `toMaterial`
- The `QTexture2D` of the settings is stored in a shared pointer.

nyalldawson commented on src/3d/materials/qgsphongtexturedmaterialsettings.h:
```diff
@@ -42,6 +43,13 @@ class _3D_EXPORT QgsPhongTexturedMaterialSettings : public QgsAbstractMaterialSe
     QgsPhongTexturedMaterialSettings() = default;
 
+    /**
+     * Constructor for QgsPhongTexturedMaterialSettings from \a phongSettings and \a texture
```
Please document what formats are acceptable for texture, and the ownership/lifetime requirements for those

ptitjano replied:
Done.

nyalldawson commented on src/3d/materials/qgstexturematerial.cpp:
```diff
@@ -0,0 +1,87 @@
+
+void QgsTextureMaterial::setTexture( Qt3DRender::QAbstractTexture *texture )
+{
+  mTextureParameter->setValue( QVariant::fromValue( texture ) );
```
What happens with ownership here?

ptitjano replied:
This is a copy/paste of upstream `Qt3DExtras::QTextureMaterial` https://doc.qt.io/qt-5/qt3dextras-qtexturematerial.html

ptitjano replied:
> This is a copy/paste of upstream `Qt3DExtras::QTextureMaterial` https://doc.qt.io/qt-5/qt3dextras-qtexturematerial.html

Ownership is transferred to the material.

nyalldawson replied:
Can you note that in the documentation please? I realise it's a copy/paste, but qt3d upstream documentation is just plain 💩 when it comes to documenting these kind of things. 😂

ptitjano replied:
Done.

nyalldawson commented on src/3d/materials/qgsphongtexturedmaterialsettings.cpp:
```diff
@@ -202,9 +246,29 @@ void QgsPhongTexturedMaterialSettings::addParametersToEffect( Qt3DRender::QEffec
+  if ( texture )
+  {
+    mDiffuseTexturePath.clear();
+    mDiffuseTexture.reset( texture );
```
I still don't get how this can be stable. QTexture2D is a QObject subclass, so we'll have both thread safety issues (if a QgsPhongTexturedMaterialSettings is copied across threads) + the risk of a parent object for the QTexture2D instance causing object deletion outside of the shared_ptr . 

@wonder-sk i'd love your insights here too, given that you're the resident expert in Qt3D internals :laughing:

wonder-sk replied:
+1 to what Nyall said... passing QTexture2D around looks quite scary (and just like it was pointed out - QObject-based class in shared_ptr is not safe).

Looking at the QgsPhongTexturedMaterialSettings::toMaterial() implementation, we end up extracting QImage from it, so why not have `QgsPhongTexturedMaterialSettings::setDiffuseTexture( const QImage &diffuse )` instead? It would simplify things, not requiring two variants (either a path or a texture). Another argument against using QTexture2D is that PyQGIS users would not be able to use that (since we do not have PyQt3D linked).

Please double check also the functionality of terrain texture updates (e.g. when a map layer in project participating in the terrain's texture images gets enabled/disabled) - in such case, we do not recreate all the entities and materials again, instead we just update the image in `QgsTerrainTextureImage`. I believe this functionality is lost in the current PR, because the material now does not use the `QgsTerrainTextureImage` object (which gets new image), so this link seems to be lost. We may need some other mechanism to make sure that terrain's texture can be updated without a complete rebuild of the material...

ptitjano replied:
> +1 to what Nyall said... passing QTexture2D around looks quite scary (and just like it was pointed out - QObject-based class in shared_ptr is not safe).

This was never my intention to use a smart pointer, that was @nyalldawson's idea :laughing: !

> 
> Looking at the QgsPhongTexturedMaterialSettings::toMaterial() implementation, we end up extracting QImage from it, so why not have `QgsPhongTexturedMaterialSettings::setDiffuseTexture( const QImage &diffuse )` instead? It would simplify things, not requiring two variants (either a path or a texture). Another argument against using QTexture2D is that PyQGIS users would not be able to use that (since we do not have PyQt3D linked).
> 

Indeed. And that's the case now. 
Thanks for your comment. I came to the same conclusion. At first, I was trying to copy/paste Qt3D behavior but It creates a Material on the fly while QGIS stores settings and then creates a Material with `toMaterial()`. When I realized that the  `QImage from the stored `QTexture2D` needs to be copied to create a new `QTexture2D`, I thought that it would be much simpler to directly use a `QImage`. 

> Please double check also the functionality of terrain texture updates (e.g. when a map layer in project participating in the terrain's texture images gets enabled/disabled) - in such case, we do not recreate all the entities and materials again, instead we just update the image in `QgsTerrainTextureImage`. I believe this functionality is lost in the current PR, because the material now does not use the `QgsTerrainTextureImage` object (which gets new image), so this link seems to be lost. We may need some other mechanism to make sure that terrain's texture can be updated without a complete rebuild of the material...

I don't understand. On current master, on terrain texture updates, `QgsTerrainTileLoader::createTextureComponent` gets called and thus the material is recreated. I don't see how only the image is updated. Based on my understanding, this is is not the case on current master.
Anyway, the terrain texture updates works on this PR.

wonder-sk replied:
Thanks for the updates!

> I don't understand. On current master, on terrain texture updates, QgsTerrainTileLoader::createTextureComponent gets called and thus the material is recreated. I don't see how only the image is updated. Based on my understanding, this is is not the case on current master.
Anyway, the terrain texture updates works on this PR.

Please try this use case:
1. open a 3D view
2. enable or disable a map layer (one that does not have a 3D renderer so that the terrain texture gets updated) - I expect this still works because QgsTextureMaterial + createTexture() are used
3. now in 3D view settings, enable terrain shading
4. enable or disable a map layer - here I am expecting that the terrain texture will not get updated anymore

The mechanism how the terrain texture is updated works like this (verified on qgis master):
- when the list of map layers for terrain texture gets modified, `QgsTerrainEntity::invalidateMapImages()` gets called
- this will create tile update jobs (`TerrainMapUpdateJob`) from `TerrainMapUpdateJobFactory` 
- when an updated texture is ready, `TerrainMapUpdateJob::onTileReady` is called and it uses `QgsTerrainTileEntity::textureImage()` to get `QgsTerrainTextureImage` and update the image

As far as I can tell, when terrain shading is enabled, `QgsTerrainTileLoader::createTextureComponent()` now does not use `createTexture()` which is the key bit here - it wraps texture's QImage into `QgsTerrainTextureImage` (which allows updates of the image without full re-creation of terrain's tile entity).

For terrain tile load, probably you may want to have an approach similar to QgsTextureMaterial::setTexture()... the `QgsPhongTexturedMaterialSettings` object would be created without any texture image, and in a second step you would pass the texture image wrapped into QgsTerrainTextureImage...

ptitjano replied:
> Thanks for the updates!
> 
> > I don't understand. On current master, on terrain texture updates, QgsTerrainTileLoader::createTextureComponent gets called and thus the material is recreated. I don't see how only the image is updated. Based on my understanding, this is is not the case on current master.
> > Anyway, the terrain texture updates works on this PR.
> 
> Please try this use case:
> 
>     1. open a 3D view
> 
>     2. enable or disable a map layer (one that does not have a 3D renderer so that the terrain texture gets updated) - I expect this still works because QgsTextureMaterial + createTexture() are used
> 
>     3. now in 3D view settings, enable terrain shading
> 
>     4. enable or disable a map layer - here I am expecting that the terrain texture will not get updated anymore
> 
> 
> The mechanism how the terrain texture is updated works like this (verified on qgis master):
> 
>     * when the list of map layers for terrain texture gets modified, `QgsTerrainEntity::invalidateMapImages()` gets called
> 
>     * this will create tile update jobs (`TerrainMapUpdateJob`) from `TerrainMapUpdateJobFactory`
> 
>     * when an updated texture is ready, `TerrainMapUpdateJob::onTileReady` is called and it uses `QgsTerrainTileEntity::textureImage()` to get `QgsTerrainTextureImage` and update the image
> 
> 
> As far as I can tell, when terrain shading is enabled, `QgsTerrainTileLoader::createTextureComponent()` now does not use `createTexture()` which is the key bit here - it wraps texture's QImage into `QgsTerrainTextureImage` (which allows updates of the image without full re-creation of terrain's tile entity).
> 
> For terrain tile load, probably you may want to have an approach similar to QgsTextureMaterial::setTexture()... the `QgsPhongTexturedMaterialSettings` object would be created without any texture image, and in a second step you would pass the texture image wrapped into QgsTerrainTextureImage...

Thanks for the explanation. I was unable to reproduce it because it only occurs when the terrain shading is enabled. 

I tried to find a solution with `QgsPhongTexturedMaterialSettings` but if we create it without a texture, it fallbacks to the non texture case which seems reasonable because we also need to handle the wrong texture path case. And we can't directly call a `setTexture` on the output of `toMaterial`. because it is merely a `QTextureMaterial`.  

So I ended up creating a new `QgsPhongTexturedMaterial` class which is mostly the code of   `QgsPhongTexturedMaterialSettings::toMaterial`.   
Now,  `QgsPhongTexturedMaterialSettings::toMaterial` calls `QgsPhongTexturedMaterial` with the good parameters and `QgsPhongTexturedMaterial` is directly used for the terrain with a `setTexture` call.    
This logic is similar to `QgsMetalRoughMaterial` / `QgsMetalRoughMaterialSettings`

ptitjano committed:
qgsterraintileloader: Fix typo

wonder-sk commented on src/3d/materials/qgsphongtexturedmaterialsettings.cpp:
```diff
@@ -104,79 +114,92 @@ Qt3DRender::QMaterial *QgsPhongTexturedMaterialSettings::toMaterial( QgsMaterial
-      const QImage textureSourceImage = QgsApplication::imageCache()->pathAsImage( mDiffuseTexturePath, QSize(), true, 1.0, fitsInCache );
-      ( void )fitsInCache;
+      Qt3DRender::QMaterial *material = nullptr;
+      QImage textureSourceImage = mDiffuseTextureImage.copy();
```
why are we doing explicit copy of the QImage here, and in QgsPhongTexturedMaterialSettings::setDiffuseTexture() ?

ptitjano committed:
qgsphongtexturedmaterialsettings: Factor out material creation

This introduces a new class `QgsPhongTexturedMaterial` which is
similar to Qt3D `QDiffuseSpecularMaterial`.
`QgsPhongTexturedMaterial::toMaterial()` now calls
`QgsPhongTexturedMaterial` to create the material. This has no
functional change.

This will make it possible to directly use `QgsPhongTexturedMaterial`
for the terrain in the next commit. Indeed, the texture case needs to
use a `QTexture2D` which cannot be stored in a settings class. This
issue is avoided by directly creating the material instead of using
`QgsPhongTexturedMaterial::toMaterial()`.

wonder-sk reviewed (approved):
:+1:  looks good to me, thanks for the updates!

ptitjano commented:
> 👍 looks good to me, thanks for the updates!

Thanks for the review.
